lib/_http_agent.js

```diff
@@ -496,7 +496,10 @@ Agent.prototype.keepSocketAlive = function keepSocketAlive(socket) {
       const hint = RegExpPrototypeExec(/^timeout=(\d+)/, keepAliveHint)?.[1];
 
       if (hint) {
-        const serverHintTimeout = NumberParseInt(hint) * 1000;
+        // Let the timer expires before the announced timeout to reduce
+        // the likelihood of ECONNRESET errors
+        let serverHintTimeout = ( NumberParseInt(hint) * 1000 ) - 1000;
+        serverHintTimeout = serverHintTimeout > 0 ? serverHintTimeout : 0;
 
         if (serverHintTimeout < agentTimeout) {
```

Comment:
This conditional needs to be fixed - agentTimeout defaults to 0, so the serverHintTimeout is never being set.

Reply:
In this case the socket shouldn't be reused at all. Is setting socket.setTimeout(0) the right way to do it?

Reply:
Take a look at [socket docs](https://nodejs.org/api/net.html#socketsettimeouttimeout-callback) - socket.setTimeout(0) doesn't mean immediate, it means never.

Reply:
Sorry, I have limited understand of the network internals. In undici ( https://github.com/nodejs/undici/pull/291/files ) if the keepAliveTimeout goes down to 0 they flag the connection as reset:

```
if (!keepAliveTimeout || keepAliveTimeout < 1e3) {
          client[kReset] = true
        } 
```

I don't know how to achieve the same result here.

Reply:
I also have a limited understanding of network internals, but I think:
```
if (!agentTimeout || serverHintTimeout < agentTimeout) {
```
is what you want.